### PR TITLE
Sonarunner 132

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: sonar-runner-impl/src/main/java/org/sonar/runner/impl/ServerConnection.java [hunk 1]
```diff
@@ -78,31 +116,23 @@ void download(String path, File toFile) {
...
     try {
-      String charset = getCharsetFromContentType(httpRequest.contentType());
-      if (charset == null || "".equals(charset)) {
-        charset = "UTF-8";
+      if(isModePreview) {
+        return wsCache.getString(serverUrl, new StringDownloader(fullUrl));
       }
-      if (!httpRequest.ok()) {
-        throw new IOException(MessageFormat.format(STATUS_RETURNED_BY_URL_IS_INVALID, fullUrl, httpRequest.code()));
+      else {
```
Comment (SonarTech) [inline sonar-runner-impl/src/main/java/org/sonar/runner/impl/ServerConnection.java]: MINOR Move this "else" on the same line that the previous closing curly brace. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: sonar-runner-impl/src/main/java/org/sonar/runner/impl/ServerConnection.java [hunk 2]
```diff
@@ -78,31 +116,23 @@ void download(String path, File toFile) {
       }
       FileUtils.deleteQuietly(toFile);
       throw new IllegalStateException("Fail to download: " + fullUrl, e);
-
     }
   }
 
-  String downloadString(String path) throws IOException {
+  String downloadStringCache(String path) throws Exception {
```
Comment (SonarTech) [inline sonar-runner-impl/src/main/java/org/sonar/runner/impl/ServerConnection.java]: MAJOR Define and throw a dedicated exception instead of using a generic one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)

File: sonar-runner-impl/src/main/java/org/sonar/runner/impl/BatchLauncher.java [hunk 3]
```diff
@@ -44,11 +48,21 @@ public BatchLauncher() {
   }
 
   public void execute(Properties props, List<Object> extensions) {
-    ServerConnection serverConnection = ServerConnection.create(props);
+    ServerConnection serverConnection = ServerConnection.create(props, getCache(props));
     JarDownloader jarDownloader = new JarDownloader(serverConnection);
     doExecute(jarDownloader, props, extensions);
   }
 
+  private PersistentCache getCache(Properties props) {
```
Comment (SonarTech) [inline sonar-runner-impl/src/main/java/org/sonar/runner/impl/BatchLauncher.java]: MAJOR Make "getCache" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java [hunk 4]
```diff
@@ -48,4 +58,48 @@ static boolean taskRequiresProject(Properties props) {
...
+  }
+
+  public static void closeQuietly(@Nullable Closeable c) {
+    if (c == null) {
+      return;
+    }
+
+    try {
+      c.close();
+    } catch (IOException e) {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java [hunk 5]
```diff
@@ -48,4 +58,48 @@ static boolean taskRequiresProject(Properties props) {
     Object task = props.get(RunnerProperties.TASK);
     return task == null || ScanProperties.SCAN_TASK.equals(task);
   }
+
+  public static void deleteQuietly(File f) {
+    try {
+      Files.walkFileTree(f.toPath(), new SimpleFileVisitor<Path>() {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java]: MAJOR Reduce this anonymous class number of lines from 25 to at most 20, or make it a named class. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1188)

File: sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java [hunk 6]
```diff
@@ -48,4 +58,48 @@ static boolean taskRequiresProject(Properties props) {
...
+          return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+          if (exc == null) {
+            Files.delete(dir);
+            return FileVisitResult.CONTINUE;
+          }
+          else {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java]: MINOR Move this "else" on the same line that the previous closing curly brace. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceSameLineAsNextBlockCheck)

File: sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java [hunk 7]
```diff
@@ -48,4 +58,48 @@ static boolean taskRequiresProject(Properties props) {
...
+            // directory iteration failed; propagate exception
+            throw exc;
+          }
+        }
+      });
+    } catch (IOException e) {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 new issues:
* 2 critical
* 3 major
* 2 minor

File: sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java [hunk 8]
```diff
@@ -48,4 +58,49 @@ static boolean taskRequiresProject(Properties props) {
     Object task = props.get(RunnerProperties.TASK);
     return task == null || ScanProperties.SCAN_TASK.equals(task);
   }
+
+  public static void deleteQuietly(File f) {
+    try {
+      Files.walkFileTree(f.toPath(), new DeleteFileVisitor());
+    } catch (IOException e) {
```
Comment (SonarTech) [inline sonar-runner-api/src/main/java/org/sonar/runner/api/Utils.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

File: sonar-runner-api/src/test/java/org/sonar/runner/api/DirsTest.java [hunk 9]
```diff
@@ -54,7 +53,7 @@ public void should_use_parameterized_task_work_dir() throws Exception {
 
     File workDir = new File(runner.property("sonar.working.directory", null));
     assertThat(workDir).isNotNull();
-    assertThat(FilenameUtils.separatorsToUnix(workDir.getCanonicalPath())).contains("generated/reports");
+    assertThat(workDir.getCanonicalPath()).contains("generated/reports");
```
Comment (henryju) [inline sonar-runner-api/src/test/java/org/sonar/runner/api/DirsTest.java]: Does it still works on Windows?

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

File: sonar-runner-impl/src/test/java/org/sonar/runner/impl/ServerConnectionTest.java [hunk 10]
```diff
@@ -38,14 +43,24 @@
   @Rule
   public TemporaryFolder temp = new TemporaryFolder();
 
+  @Rule
+  public TemporaryFolder cacheTmp = new TemporaryFolder();
```
Comment (henryju) [inline sonar-runner-impl/src/test/java/org/sonar/runner/impl/ServerConnectionTest.java]: Why did you add a new TemporaryFolder @Rule? There is already one 2 lines before :)

**Comment by henryju** [comment]
LGTM appart two minor comments

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 2 critical
* 1 major
